### PR TITLE
[JENKINS-58101] Cache blockage reasons when considering parked executors

### PR DESCRIPTION
**Pull request by res0nance**
See [JENKINS-58101](https://issues.jenkins-ci.org/browse/JENKINS-58101).

Looking to reduce the lock contention on the queue.
Perhaps we should cache it for the entire loop https://github.com/jenkinsci/jenkins/blob/a36ac5f6f6da8c2cd161672becf40ae7c9830037/core/src/main/java/hudson/model/Queue.java#L1607-L1685
But the cause of blockage does change between items due to https://github.com/jenkinsci/jenkins/blob/a36ac5f6f6da8c2cd161672becf40ae7c9830037/core/src/main/java/hudson/model/Queue.java#L277

### Proposed changelog entries

* Entry 1: JENKINS-58101, Reduce lock contention around jenkins queue
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/model/Queue.java [hunk 1]
```diff
@@ -1631,17 +1632,23 @@ public void maintain() {
...
                     for (JobOffer j : parked.values()) {
-                        CauseOfBlockage reason = j.getCauseOfBlockage(p);
+                        Node offerNode = j.getNode();
+                        CauseOfBlockage reason;
+                        if (reasonMap.containsKey(offerNode)) {
+                            reason = reasonMap.get(offerNode);
+                        } else {
+                            reason = j.getCauseOfBlockage(p);
+                            reasonMap.put(offerNode, reason);
+                        }
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/Queue.java:1644]: Putting null to reasonMap is expected IIUC

```suggestion
                        CauseOfBlockage reason = reasonMap.get(offerNode);
                        if (reason == null) {
                            reason = j.getCauseOfBlockage(p);
                            reasonMap.put(offerNode, reason);
                        }
```

**Review by oleg-nenashev — APPROVED** [review]
Just a refactoring proposal

File: core/src/main/java/hudson/model/Queue.java (hunk 1 above)
Comment (res0nance) [reply core/src/main/java/hudson/model/Queue.java:1644]: Its ok to have nulls as values as the reason can be null so the suggested version will call getCauseOfBlockage more than once

Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/Queue.java:1644]: OK

**Comment by oleg-nenashev** [comment]
@jenkinsci/core additional reviews would be appreciated

**Review by timja — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Comment by oleg-nenashev** [comment]
Thanks @res0nance !

**Pull request merged**